server/services/store/sqlstore/blocks.go

```diff
@@ -788,3 +802,244 @@ func (s *SQLStore) duplicateBlock(db sq.BaseRunner, boardID string, blockID stri
+		WHERE bh.id = sub.id AND bh.insert_at=sub.max_insert_at AND bh.delete_at <> 0
+	*/
+	selectQuery := s.getQueryBuilder(db).
+		Select(
+			"bh.board_id",
+			"'' AS channel_id",
+			"bh.id",
+			"bh.parent_id",
+			"bh.schema",
+			"bh.type",
+			"bh.title",
+			"bh.fields",
+			"'"+modifiedBy+"' AS modified_by",
+			"bh.create_at",
+			s.castInt(utils.GetMillis(), "update_at"),
+			s.castInt(0, "delete_at"),
+			"bh.created_by",
+		).
+		From(fmt.Sprintf(`
+				%sblocks_history AS bh,
+				(SELECT id, max(insert_at) AS max_insert_at FROM %sblocks_history WHERE %s GROUP BY id) AS sub`,
+			s.tablePrefix, s.tablePrefix, where)).
+		Where("bh.id=sub.id").
+		Where("bh.insert_at=sub.max_insert_at").
+		Where(sq.NotEq{"bh.delete_at": 0})
+
+	columns := []string{
+		"board_id",
+		"channel_id",
+		"id",
+		"parent_id",
+		s.escapeField("schema"),
+		"type",
+		"title",
+		"fields",
+		"modified_by",
+		"create_at",
+		"update_at",
+		"delete_at",
+		"created_by",
+	}
+
+	insertQuery := s.getQueryBuilder(db).Insert(s.tablePrefix + "blocks").
+		Columns(columns...).
+		Select(selectQuery)
+
+	insertHistoryQuery := s.getQueryBuilder(db).Insert(s.tablePrefix + "blocks_history").
+		Columns(columns...).
+		Select(selectQuery)
+
+	sql, args, err := insertQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	sql, args, err = insertHistoryQuery.ToSql()
+	s.logger.Trace("undeleteBlockChildren - insertHistoryQuery",
+		mlog.String("sql", sql),
+		mlog.Array("args", args),
+		mlog.Err(err),
+	)
+
+	// insert into blocks table must happen before history table, otherwise the history
+	// table will be changed and the second query will fail to find the same records.
+	result, err := insertQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ := result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	result, err = insertHistoryQuery.Exec()
+	if err != nil {
+		return err
+	}
+	rowsAffected, _ = result.RowsAffected()
+	s.logger.Debug("undeleteBlockChildren - insertHistoryQuery", mlog.Int64("rows_affected", rowsAffected))
+
+	return nil
+}
+
+func (s *SQLStore) findOrphansForBoards(db sq.BaseRunner, limit int) ([]string, error) {
+	if limit <= 0 {
+		limit = 1000
+	}
+	if limit > 1000 {
+		return nil, ErrLimitExceeded{1000}
+	}
+
+	// Find all orphaned child blocks for deleted boards.
+	// Squirrel can't express multiple tables in FROM, nor can it mix `From` and `FromSelect`.
+	sql := `
+		SELECT fb.id from %sblocks AS fb WHERE fb.board_id IN
+		(
+			SELECT bh.id FROM %sboards_history AS bh,
+				(SELECT id, max(insert_at) AS max_insert_at FROM %sboards_history GROUP BY id) AS sub
+			WHERE bh.id=sub.id AND bh.insert_at=sub.max_insert_at AND bh.delete_at > 0
+		)
+		order by fb.insert_at desc limit %d;
+	`
+	sql = fmt.Sprintf(sql, s.tablePrefix, s.tablePrefix, s.tablePrefix, limit)
```

Comment:
> It shouldn't be a problem, because you are concatenating using a `%d` there, but I would prefer to put an `?` and pass the `limit` as a parameter to the `db.Query`

There is no possibility for injection here, but you're right, it's a bad example for other developers. I'll fix.

Reply:
> you said you can't use the multiple from columns in the squirrel select, but looks like you are using it in the previous method

TBH, I think the squirrel version of these particular queries are far less readable and harder to construct than straight SQL.  I used squirrel in the previous query because there were multiple queries needed which I composed as sections with a large column list as well.  I had to stuff a bunch of stuff into the `From(...)` API that doesn't belong there.  Overall the economy of re-using the sub-queries may be worth the less readable code. :man_shrugging: 

> but I would prefer to put an ? and pass the limit as a parameter

Unfortunately we cannot use placeholders for table names, so I still need to compose the string for those. I did modify it to pass the limit.

Reply:
I'm ok with that for table name with the tablePrefix as input, because that is not an "external" input. But yes, still an example that is far from be perfect.

The reason for that, is to allow on-prem or even allow to reuse your same database adding a prefix for the table name. Probably we can get rid of that in the mid term when we are only the boards product.